plugin/output/postgres/postgres.go

```diff
@@ -0,0 +1,348 @@
+
+	db, err := pgx.Connect(pgx.ConnConfig{
+		Host:     p.config.Host,
+		Port:     p.config.Port,
+		Database: p.config.DBname,
+		Password: p.config.Pwd,
+		User:     p.config.Username,
+		// to work with
+		PreferSimpleProtocol: true,
+	})
+	if err != nil {
+		p.logger.Fatalf("can't connect to pgsql: %v", err)
+	}
+	p.db = db
+
+	p.batcher = pipeline.NewBatcher(
+		params.PipelineName,
+		outPluginType,
+		p.out,
+		nil,
+		p.controller,
+		p.config.WorkersCount_,
+		p.config.BatchSize_,
+		p.config.BatchFlushTimeout_,
+		0,
+	)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	p.ctx = ctx
+	p.cancelFunc = cancel
+
+	p.batcher.Start(ctx)
+}
+
+func (p *Plugin) Stop() {
+	p.cancelFunc()
+	p.batcher.Stop()
+}
+
+func (p *Plugin) Out(event *pipeline.Event) {
+	p.batcher.Add(event)
+}
+
+func (p *Plugin) out(workerData *pipeline.WorkerData, batch *pipeline.Batch) {
+	builder := p.queryBuilder.GetInsertBuilder()
+	pgFields := p.queryBuilder.GetPgFields()
+	uniqueFields := p.queryBuilder.GetUniqueFields()
+
+	// Deduplicate events, pg can't do upsert with duplication.
+	uniqueEventsMap := make(map[string]struct{}, len(batch.Events))
+	var atLeastOneGoodEvent bool
+
+	for _, event := range batch.Events {
+		lVals := make([]interface{}, 0, len(pgFields))
+		uniqueID := ""
+
+	LOOP_THROUGH_EVENT:
+		for _, field := range pgFields {
+			fNode, err := event.Root.DigStrict(field.Name)
+			if err != nil {
+				// In strict mode file.d fails, otherwise truncate bad event.
+				if p.config.Strict {
+					p.logger.Fatalf("Required field '%s' doesn't appear in event: %s", field.Name, string(event.Buf))
+				}
+
+				p.logger.Error("Required field '%s' doesn't appear in event: %s", field.Name, string(event.Buf))
+				stats.GetCounter(subsystemName, discardedEventCounter).Inc()
+				// stop loop for event
+				break LOOP_THROUGH_EVENT
+			}
+
+			lVal, err := p.addFieldToValues(field, fNode)
+			if err != nil {
+				if p.config.Strict {
+					p.logger.Fatal(err)
+				}
+				p.logger.Error(err)
+				stats.GetCounter(subsystemName, discardedEventCounter).Inc()
+				break LOOP_THROUGH_EVENT
+			}
+
+			lVals = append(lVals, lVal)
+
+			if _, ok := uniqueFields[field.Name]; ok {
+				if field.ColType == pgInt || field.ColType == pgTimestamp {
+					uniqueID += strconv.Itoa(lVal.(int))
+				} else {
+					uniqueID += lVal.(string)
+				}
+			}
+		}
+
+		// If LOOP_THROUGH_EVENT was breaked values mustn't be added.
+		if len(lVals) == len(pgFields) {
+			if _, ok := uniqueEventsMap[uniqueID]; ok {
+				stats.GetCounter(subsystemName, duplicatedEventCounter).Inc()
+				p.logger.Infof("event duplicated. Fields: %v, values: %v", pgFields, lVals)
+			} else {
+				uniqueEventsMap[uniqueID] = struct{}{}
+				atLeastOneGoodEvent = true
+				builder = builder.Values(lVals...)
+			}
+		}
+	}
+
+	builder = builder.Suffix(p.queryBuilder.GetPostfix()).PlaceholderFormat(sq.Dollar)
+
+	// no valid events passed in non-strict mode.
+	if !atLeastOneGoodEvent {
```

Comment:
idea: maybe replace atLeastOneGoodEvent with haveNewEvents or smth similar

Reply:
It's not about new events, but about event, that valid occur to config: events without required fields or with wrong type are invalid.
I'll rename it to `haveValidEvents`